lib/rzo/app/config_validation.rb

```diff
@@ -0,0 +1,144 @@
+      def validate_safety_checks!(config)
+        ctx = self
+        issues = checks.each_with_object([]) do |mth, ary|
+          debug "Checking config for #{mth} issues"
+          if issue = ctx.send(mth, config)
+            ary << issue
+          end
+        end
+        if issues.empty?
+          debug 'No issues detected with the config file.'
+        else
+          validate_inform!(issues)
+        end
+      end
+      # rubocop:enable Metrics/MethodLength
+
+      ##
+      # Validate the configuration has a top level key named "defaults" and the
+      # value is a Hash map.
+      # rubocop:disable Metrics/MethodLength
+      def validate_defaults_key(config)
+        if defaults = config['defaults']
+          return Issue.new('Top level key "defaults" must have a Hash value') unless defaults.is_a? Hash
+        else
+          return Issue.new('Configuration does not contain top level "defaults" key')
+        end
+        if pth = defaults['bootstrap_repo_path']
+          return Issue.new('defaults.bootstrap_repo_path is not a String') unless pth.is_a? String
+        else
+          return Issue.new 'Configuration "defaults" value does not contain a '\
+            '"bootstrap_repo_path" key.  For example, '\
+            '{"defaults":{"bootstrap_repo_path":"/tmp/foo"}}'
+        end
+      end
+      # rubocop:enable Metrics/MethodLength
+
+      ##
+      # Validate the top level "control_repos" key, which should have a value of
+      # Array<String> where each string value is a fully qualified path.
+      def validate_control_repos(config)
+        if repos = config['control_repos']
+          return Issue.new('Top level key "control_repos" must have an Array value') unless repos.is_a? Array
+        else
+          return Issue.new('Top level key "control_repos" is not specified.  It must be an Array of paths to your control repos.')
+        end
+        repos.each do |pth|
+          if issue = validate_existence(pth)
+            return issue
+          end
+        end
+      end
+
+      ##
+      # Given a string, validate it's a fully qualified path, readable, and a
+      # git directory.
+      #
+      # Return nil if all validations pass, Otherwise return the Issue found.
+      def validate_existence(path)
+        pn = Pathname.new(path)
+        git = pn + '.git'
+        return Issue.new("#{pn} is not an absolute path.  It must be fully qualified, not relative") unless pn.absolute?
+        return Issue.new("#{pn} is not a directory.  Has it been cloned?") unless pn.directory?
+        return Issue.new("#{pn} is not readable.  Are permissions correct?") unless pn.readable?
+        return Issue.new "#{git} does not exist.  Has #{git.dirname} been coloned properly?" unless git.directory?
+      end
+
+      ##
+      # Validate the bootstrap repo path exists.
+      def validate_bootstrap_repo_path(config)
+        validate_existence(config['defaults']['bootstrap_repo_path'])
```

Comment:
perhaps we should define an array of keys to check with validate_existence?

Reply:
Good idea, get's a little tricky since they're nested but I'll definitely look into it.